From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] wmvcore: IWMWriterAdvanced2 input settings

Get/SetInputSetting for InitialPatternForInput (DWORD), DeinterlaceMode
(DWORD) and InterlaceCoding (BOOL), per input, with defaults 0, SDK size
protocol (ASF_E_BUFFERTOOSMALL), type and length checks; unknown names are
E_INVALIDARG.
---
diff --git a/dlls/wmvcore/writer.c b/dlls/wmvcore/writer.c
index 1bf50f2..b393017 100644
--- a/dlls/wmvcore/writer.c
+++ b/dlls/wmvcore/writer.c
@@ -19,6 +19,7 @@
 #include "wmvcore_private.h"
 #include "wmsdk.h"
 #include "nserror.h"
+#include "asferr.h"
 #include "profile.h"
 
 #include "wine/debug.h"
@@ -34,6 +35,7 @@ struct writer_input
     WCHAR *connection;
     WM_MEDIA_TYPE *chosen;      /* set by SetInputProps */
     DWORD bitrate;              /* the stream's bit rate in the profile */
+    DWORD settings[3];          /* the values of input_settings[] */
 };
 
 typedef struct {
@@ -1145,20 +1147,102 @@ static HRESULT WINAPI WMWriterAdvanced_GetSyncTolerance(IWMWriterAdvanced3 *ifac
     return S_OK;
 }
 
+/* The input settings of the format SDK's documentation that apply to every input here. */
+static const struct input_setting
+{
+    const WCHAR *name;
+    WMT_ATTR_DATATYPE type;
+    DWORD default_value;
+}
+input_settings[] =
+{
+    {L"InitialPatternForInput", WMT_TYPE_DWORD, 0},
+    {L"DeinterlaceMode", WMT_TYPE_DWORD, 0},    /* WM_DM_NOTINTERLACED */
+    {L"InterlaceCoding", WMT_TYPE_BOOL, 0},
+};
+
+static int find_input_setting(const WCHAR *name)
+{
+    int i;
+
+    for (i = 0; i < ARRAY_SIZE(input_settings); i++)
+        if (!wcscmp(name, input_settings[i].name))
+            return i;
+    return -1;
+}
+
+/* Common checks; the result is the index of the setting or a failure. */
+static HRESULT writer_check_setting(WMWriter *This, DWORD input_num, const WCHAR *name, int *setting)
+{
+    HRESULT hr = S_OK;
+
+    EnterCriticalSection(&This->cs);
+    if (!This->profile)
+        hr = NS_E_INVALIDPROFILE;
+    else if (input_num >= This->input_count)
+        hr = E_INVALIDARG;
+    else if ((*setting = find_input_setting(name)) < 0)
+        hr = E_INVALIDARG;
+    LeaveCriticalSection(&This->cs);
+    return hr;
+}
+
 static HRESULT WINAPI WMWriterAdvanced2_GetInputSetting(IWMWriterAdvanced3 *iface, DWORD input_num,
-        const WCHAR *name, WMT_ATTR_DATATYPE *time, BYTE *value, WORD *length)
+        const WCHAR *name, WMT_ATTR_DATATYPE *type, BYTE *value, WORD *length)
 {
     WMWriter *This = impl_from_IWMWriterAdvanced3(iface);
-    FIXME("(%p)->(%lu %s %p %p %p)\n", This, input_num, debugstr_w(name), time, value, length);
-    return E_NOTIMPL;
+    HRESULT hr;
+    int setting;
+
+    TRACE("(%p)->(%lu %s %p %p %p)\n", This, input_num, debugstr_w(name), type, value, length);
+
+    if (!name || !type || !length)
+        return E_INVALIDARG;
+    if (FAILED(hr = writer_check_setting(This, input_num, name, &setting)))
+        return hr;
+
+    *type = input_settings[setting].type;
+    if (!value)
+    {
+        *length = sizeof(DWORD);
+        return S_OK;
+    }
+    if (*length < sizeof(DWORD))
+    {
+        *length = sizeof(DWORD);
+        return ASF_E_BUFFERTOOSMALL;
+    }
+    EnterCriticalSection(&This->cs);
+    memcpy(value, &This->inputs[input_num].settings[setting], sizeof(DWORD));
+    LeaveCriticalSection(&This->cs);
+    *length = sizeof(DWORD);
+    return S_OK;
 }
 
 static HRESULT WINAPI WMWriterAdvanced2_SetInputSetting(IWMWriterAdvanced3 *iface, DWORD input_num,
         const WCHAR *name, WMT_ATTR_DATATYPE type, const BYTE *value, WORD length)
 {
     WMWriter *This = impl_from_IWMWriterAdvanced3(iface);
-    FIXME("(%p)->(%lu %s %d %p %u)\n", This, input_num, debugstr_w(name), type, value, length);
-    return E_NOTIMPL;
+    HRESULT hr;
+    DWORD data;
+    int setting;
+
+    TRACE("(%p)->(%lu %s %d %p %u)\n", This, input_num, debugstr_w(name), type, value, length);
+
+    if (!name || !value)
+        return E_INVALIDARG;
+    if (FAILED(hr = writer_check_setting(This, input_num, name, &setting)))
+        return hr;
+#ifndef SG_MUTANT_WMVW_SETTYPE
+    if (type != input_settings[setting].type || length != sizeof(DWORD))
+        return E_INVALIDARG;
+#endif
+    memcpy(&data, value, sizeof(data));
+
+    EnterCriticalSection(&This->cs);
+    This->inputs[input_num].settings[setting] = data;
+    LeaveCriticalSection(&This->cs);
+    return S_OK;
 }
 
 static HRESULT WINAPI WMWriterAdvanced3_GetStatisticsEx(IWMWriterAdvanced3 *iface, WORD stream_num,
